kernel/device.c

```diff
@@ -23,6 +23,8 @@ extern const struct init_entry __init_SMP_start[];
 extern struct device __device_start[];
 extern struct device __device_end[];
 
+extern struct device_context __device_context_start[];
```

Comment:
This approach seems to disallow a specific driver implementation from extending the base context with its own additional context.  Having `struct device_context *context` in `struct_device` would be more general, allowing `CONTAINER_OF` to upcast to access additional information.

This could complicate the initialization macros; OTOH the context could be put into BSS and the sync/lock fields initialized on boot by the init loop.

Reply:
struct device's data attribute is the device driver dedicated context.
I thought we could merge both, as I did for dedicated api exposure in net stack for 15.4 (othe l2 have been using the same trick then). All specialized api have to get a net_if_api attribute in first place.

But it could be done afterwards (or before, either way), in order not to do too much changes in this PR. It's somehow 2 different things. Thus also why I did not add yet another pointer in struct device.